### PR TITLE
Implement a groupchat accept screen.

### PR DESCRIPTION
**Pull request by robinlinden**
I don't remember where I was on this, but it's not done.

---
This change is [Reviewable](https://reviewable.io/reviews/utox/utox/1074)

### CODE REVIEW

**Comment by GrayHatter** [comment]
Reviewed 3 of 15 files at r1, 26 of 38 files at r2.
Review status: 29 of 47 files reviewed at latest revision, 2 unresolved discussions.

---

*[langs/en.h, line 338 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYUwFjdZn8mbio5Zdg:-KvYUwFkWN7r57DdEUsC:blp6ggp) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/langs/en.h#L338)):*
> ```objc
> 
> msgid(GROUP_INVITE_FRIEND)
> msgstr("%s invites you to the chat")
> ```

we don't allow null termination from data that could come from untrusted users.

---

*[src/layout/group.c, line 65 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYVPWIVM2VcowGYavH:-KvYVPWIVM2VcowGYavI:ba6gr9g) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/src/layout/group.c#L65)):*
> ```C
>             }
> 
>             unsigned w = textwidth(buf, text_length);
> ```

this can overrun.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by redmanmale** [comment]
Review status: 27 of 47 files reviewed at latest revision, 2 unresolved discussions.

---

*[langs/en.h, line 338 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYUwFjdZn8mbio5Zdg:-KvYsLnRbb2vlvc9bKPY:b-896fix) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/langs/en.h#L338)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

we don't allow null termination from data that could come from untrusted users.
</blockquote></details>

Done.

---

*[src/layout/group.c, line 65 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYVPWIVM2VcowGYavH:-KvYsPJqTiIwEsNJ9_1p:b-896fix) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/src/layout/group.c#L65)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

this can overrun.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by robinlinden** [comment]
Reviewed 8 of 15 files at r1, 32 of 38 files at r2, 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by robinlinden** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[langs/en.h, line 335 at r3](https://reviewable.io:443/reviews/utox/utox/1074#-Kvs47_5TM15ZAiD-yEV:-Kvs47_5TM15ZAiD-yEW:bpzpnyu) ([raw file](https://github.com/utox/utox/blob/a391d053acd1582732d7dcb6a3f7bfe92722a72c/langs/en.h#L335)):*
> ```objc
> 
> msgid(GROUP_INVITE)
> msgstr("Chat invite")
> ```

I'm not a fan of the change from groupchat to chat. Is this something we want to do?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by robinlinden** [comment]
Reviewed 18 of 18 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by robinlinden** [comment]
Reviewed 1 of 15 files at r1, 1 of 38 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/layout/group_invite.c, line 23 at r4](https://reviewable.io:443/reviews/utox/utox/1074#-KyD_6D8uuVO_WCsZuMX:-KyD_6D8uuVO_WCsZuMY:byra413) ([raw file](https://github.com/utox/utox/blob/4252d7130711411f5cd6da538587e7906cbf5bb7/src/layout/group_invite.c#L23)):*
> ```C
> static void button_group_invite_reject_on_mup(void);
> 
> void group_invite_draw(void);
> ```

This declaration isn't needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by redmanmale** [comment]
Review status: 46 of 47 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/layout/group_invite.c, line 23 at r4](https://reviewable.io:443/reviews/utox/utox/1074#-KyD_6D8uuVO_WCsZuMX:-KyFs4gDeGjiFPaNCnW-:b-896fix) ([raw file](https://github.com/utox/utox/blob/4252d7130711411f5cd6da538587e7906cbf5bb7/src/layout/group_invite.c#L23)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

This declaration isn't needed.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by robinlinden** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by endoffile78** [comment]
Reviewed 8 of 15 files at r1, 18 of 38 files at r2, 2 of 7 files at r3, 18 of 18 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/group_invite.c, line 30 at r5](https://reviewable.io:443/reviews/utox/utox/1074#-L-cjR7S_Nzwn8CLMrZ4:-L-cjR7TG3fX5sa1A-UV:b13m6zj) ([raw file](https://github.com/utox/utox/blob/56b05a3653909b67f11e9cc56cad30410ee99133/src/group_invite.c#L30)):*
> ```C
> }
> 
> static void group_invite_free(const uint8_t invite_id) {
> ```

Add a check to make sure `invite_id` is less than 16

---

*[src/group_invite.c, line 123 at r5](https://reviewable.io:443/reviews/utox/utox/1074#-L-cjZZUpz2eIM4qmoeG:-L-cjZZUpz2eIM4qmoeH:b13m6zj) ([raw file](https://github.com/utox/utox/blob/56b05a3653909b67f11e9cc56cad30410ee99133/src/group_invite.c#L123)):*
> ```C
> }
> 
> uint32_t group_invite_get_friend_id(const uint8_t invite_id) {
> ```

Add a check to make sure `invite_id` is less than 16

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by im-grey** [comment]
Thank you for all your hard work.

**Comment by redmanmale** [comment]
Review status: 32 of 48 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/group_invite.c, line 30 at r5](https://reviewable.io:443/reviews/utox/utox/1074#-L-cjR7S_Nzwn8CLMrZ4:-L-gmix4pLDCdqGM2Pdt:b-896fix) ([raw file](https://github.com/utox/utox/blob/56b05a3653909b67f11e9cc56cad30410ee99133/src/group_invite.c#L30)):*
<details><summary><i>Previously, endoffile78 (Endoffile) wrote…</i></summary><blockquote>

Add a check to make sure `invite_id` is less than 16
</blockquote></details>

Done.

---

*[src/group_invite.c, line 123 at r5](https://reviewable.io:443/reviews/utox/utox/1074#-L-cjZZUpz2eIM4qmoeG:-L-gmk9JBWvqo0_omaob:b-896fix) ([raw file](https://github.com/utox/utox/blob/56b05a3653909b67f11e9cc56cad30410ee99133/src/group_invite.c#L123)):*
<details><summary><i>Previously, endoffile78 (Endoffile) wrote…</i></summary><blockquote>

Add a check to make sure `invite_id` is less than 16
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by endoffile78** [comment]
:lgtm:

---

Reviewed 14 of 14 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074#-:-L-goT46XPyNe_bHBebo:bnfp4nl)*

**Comment by GrayHatter** [comment]
"TOUCH ALL THE FILES" -- @robinlinden probably 

round 1 review

---

Reviewed 2 of 15 files at r1, 13 of 18 files at r4, 8 of 14 files at r6, 1 of 2 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[langs/en.h, line 335 at r3](https://reviewable.io:443/reviews/utox/utox/1074#-Kvs47_5TM15ZAiD-yEV:-L0kGl5gr0TX0qTZBsXK:b-d7j3ej) ([raw file](https://github.com/utox/utox/blob/a391d053acd1582732d7dcb6a3f7bfe92722a72c/langs/en.h#L335)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

I'm not a fan of the change from groupchat to chat. Is this something we want to do?
</blockquote></details>

Not in this pull, but we should change everything to conference. Per the toxcore change.

Then groups, and conferences can live side by side. Both with a different feature set.

---

*[langs/i18n_decls.h, line 107 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYVxeGUvq0BuciG1Hk:-KvYVxeGUvq0BuciG1Hl:bz7tcar) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/langs/i18n_decls.h#L107)):*
> ```objc
>     STR_GROUP_MESSAGE_INVITE,
>     STR_GROUP_MESSAGE_JOIN,
>     STR_GROUP_MESSAGE_CHANGE_NAME,
> ```

Do we want to add a suffix for strings that required additional params?

@robinlinden 

---

*[src/ui.c, line 314 at r7](https://reviewable.io:443/reviews/utox/utox/1074#-L0kHHvSOEguykPcIlPf:-L0kHHvSOEguykPcIlPg:bngpwbl) ([raw file](https://github.com/utox/utox/blob/da2788596a1bb3f80e8aa204864b7e521592423d/src/ui.c#L314)):*
> ```C
>     /* Group Settings */
>     CREATE_EDIT(group_topic, 10, 95, -10, 24);
>     CREATE_DROPDOWN(notify_groupchats, 10, 150, 24, 100);
> ```

We don't use this macro anymore. Move it to the UI file.

---

*[src/ui.c, line 344 at r7](https://reviewable.io:443/reviews/utox/utox/1074#-L0kHWKIpvW4DNic7YRW:-L0kHWKIpvW4DNic7YRX:bn8x6bd) ([raw file](https://github.com/utox/utox/blob/da2788596a1bb3f80e8aa204864b7e521592423d/src/ui.c#L344)):*
> ```C
>     CREATE_BUTTON(deny_deletion,    110, MAIN_TOP + 40, _BM_SBUTTON_WIDTH, _BM_SBUTTON_HEIGHT);
> 
>     group_invite_draw();
> ```

NTS: what is this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074#-:-L0kGa07j-gyD-NfT7I2:b-c1uiz1)*

**Comment by redmanmale** [comment]
Review status: 46 of 48 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/ui.c, line 314 at r7](https://reviewable.io:443/reviews/utox/utox/1074#-L0kHHvSOEguykPcIlPf:-L0kXE6Pca6TUt8nmg7w:b-3ew0qp) ([raw file](https://github.com/utox/utox/blob/da2788596a1bb3f80e8aa204864b7e521592423d/src/ui.c#L314)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

We don't use this macro anymore. Move it to the UI file.
</blockquote></details>

This whole commit moved to #1079.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by endoffile78** [comment]
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by GrayHatter** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/layout/group.c, line 65 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYVPWIVM2VcowGYavH:-L3dkVab8Gth5pNbQv35:b6zq1lm) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/src/layout/group.c#L65)):*
<details><summary><i>Previously, redmanmale (redmanmale) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Sorry, this is gonna take me some time to review. This being safe depends on how each platform searches string length, + where/how each func counts bytes, vs utf8-chars.

First glance your code looks correct, but I need to make sure all the other functions are being used correctly as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by redmanmale** [comment]
Reviewed 3 of 15 files at r1, 13 of 38 files at r2, 1 of 7 files at r3, 13 of 18 files at r4, 1 of 1 files at r5, 13 of 14 files at r6, 2 of 2 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/layout/group.c, line 65 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYVPWIVM2VcowGYavH:-L3e1F-hdKVS6t7wC_AO:b-pyevcy) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/src/layout/group.c#L65)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

Sorry, this is gonna take me some time to review. This being safe depends on how each platform searches string length, + where/how each func counts bytes, vs utf8-chars.

First glance your code looks correct, but I need to make sure all the other functions are being used correctly as well.
</blockquote></details>

OK, but I did'n touch this at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by GrayHatter** [comment]
@redmanmale You're right, but I can't LGTM a section that might be wrong. Hence the apology, through no fault of your own, I have to make you wait for code that's likely fine.

**Comment by redmanmale** [comment]
@GrayHatter 
Looks like it's what you are afraid of.
https://github.com/uTox/uTox/issues/1165#issuecomment-360961915

**Comment by GrayHatter** [comment]
*[src/layout/group.c, line 65 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYVPWIVM2VcowGYavH:-L4gQRJXPyQZ25bqAykV:baoexqu) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/src/layout/group.c#L65)):*
<details><summary><i>Previously, redmanmale (redmanmale) wrote…</i></summary><blockquote>

OK, but I did'n touch this at all.
</blockquote></details>

This is unsafe (not really your fault), windows uses a wchar buffer in utf8tonative()

It's likely broken throughout uTox... see `src/windows/drawing.c@textwidth`

@robinlinden @endoffile78 @cebe 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by GrayHatter** [comment]
*[src/layout/group.c, line 65 at r2](https://reviewable.io:443/reviews/utox/utox/1074#-KvYVPWIVM2VcowGYavH:-L4gRjINa5lJZgqVC6tH:b-2c99h6) ([raw file](https://github.com/utox/utox/blob/c4ca0903c11cc72142056f9a8eee9fb8fea62213/src/layout/group.c#L65)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

This is unsafe (not really your fault), windows uses a wchar buffer in utf8tonative()

It's likely broken throughout uTox... see `src/windows/drawing.c@textwidth`

@robinlinden @endoffile78 @cebe 
</blockquote></details>

LGTM 

per IRC... I'm an idiot. I confused the xlib, and windows implementations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by GrayHatter** [comment]
Reviewed 1 of 15 files at r1, 1 of 38 files at r2, 1 of 7 files at r3, 1 of 1 files at r5, 3 of 14 files at r6, 1 of 2 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/messages.c, line 194 at r8](https://reviewable.io:443/reviews/utox/utox/1074#-L4gUA26dbquGTL6ks4W:-L4gUA26dbquGTL6ks4X:bdoploq) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/messages.c#L194)):*
> ```C
>         GROUPCHAT *g = flist_get_groupchat();
> 
>         if (g && g == get_group(m->id)) {
> ```

what is this section trying to do?

---

*[src/layout/group_invite.c, line 27 at r8](https://reviewable.io:443/reviews/utox/utox/1074#-L4gSRjRSZuriP2JLcug:-L4gSRjRSZuriP2JLcuh:br0ne9w) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/layout/group_invite.c#L27)):*
> ```C
>     .button_text = { .i18nal = STR_ACCEPT },
>     .on_mup      = button_group_invite_accept_on_mup,
>     .update      = button_setcolors_success,
> ```

how often is `.update` called?

---

*[src/layout/group_invite.c, line 88 at r8](https://reviewable.io:443/reviews/utox/utox/1074#-L4gSiSTXYcWtJrGfpts:-L4gSiSTXYcWtJrGfptt:b-a234cw) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/layout/group_invite.c#L88)):*
> ```C
> 
> void group_invite_draw(void) {
>     CREATE_BUTTON(group_invite_accept, -MAIN_TOP, -80, _BM_SBUTTON_WIDTH, _BM_SBUTTON_HEIGHT);
> ```

add the positions to the structs above, and this doesn't' need to exist. (The goal is to remove this MACRO)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by redmanmale** [comment]
Review status: 40 of 47 files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[src/messages.c, line 194 at r8](https://reviewable.io:443/reviews/utox/utox/1074#-L4gUA26dbquGTL6ks4W:-L4zUPuJLtHo2idn_4RB:b-m0txxj) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/messages.c#L194)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

what is this section trying to do?
</blockquote></details>

Getting current group to scroll its content; if you're asking why I touched this at all it's because old code (`flist_get_groupchat() == get_group(m->id)`) was causing a SEGV and I fixed it.

---

*[src/layout/group_invite.c, line 27 at r8](https://reviewable.io:443/reviews/utox/utox/1074#-L4gSRjRSZuriP2JLcug:-L4zU5eK6wey5Rt-lglx:b-3twhaw) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/layout/group_invite.c#L27)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

how often is `.update` called?
</blockquote></details>

On every UI event; but we do this thing (`update = button_setcolors_success`) pretty often.

---

*[src/layout/group_invite.c, line 88 at r8](https://reviewable.io:443/reviews/utox/utox/1074#-L4gSiSTXYcWtJrGfpts:-L4zU3pVOsd0UEKq0VuG:b-896fix) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/layout/group_invite.c#L88)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

add the positions to the structs above, and this doesn't' need to exist. (The goal is to remove this MACRO)
</blockquote></details>

Done.

---

*[src/ui.c, line 344 at r7](https://reviewable.io:443/reviews/utox/utox/1074#-L0kHWKIpvW4DNic7YRW:-L4zVE1GCVItwuMxoCwe:bbg8j8u) ([raw file](https://github.com/utox/utox/blob/da2788596a1bb3f80e8aa204864b7e521592423d/src/ui.c#L344)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

NTS: what is this?
</blockquote></details>

Done. Moved to UI definition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by GrayHatter** [comment]
Reviewed 1 of 14 files at r6, 3 of 4 files at r9, 3 of 5 files at r10.
Review status: 45 of 47 files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[src/layout/group_invite.c, line 27 at r8](https://reviewable.io:443/reviews/utox/utox/1074#-L4gSRjRSZuriP2JLcug:-L6Ak_x1QNUtXDMeO5DK:b-rasr3m) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/layout/group_invite.c#L27)):*
<details><summary><i>Previously, redmanmale (redmanmale) wrote…</i></summary><blockquote>

On every UI event; but we do this thing (`update = button_setcolors_success`) pretty often.
</blockquote></details>

LGTM but I'd also like to see `// TODO: OPTIMIZE` comments when we're doing something the easy way that is a good target for micro optimization.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/utox/utox/1074)*

**Comment by redmanmale** [comment]
Review status: 18 files need 2 more reviewers, 3 open discussions remaining, 1 assignee LGTMs are missing, missing Admin LGTM at current revision. (waiting on @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, and @undefined)

---

*[src/layout/group_invite.c, line 27 at r8](https://reviewable.io/reviews/utox/utox/1074#-L4gSRjRSZuriP2JLcug:-LF7eCog8bSXnpQTLsp-:b-896fix) ([raw file](https://github.com/utox/utox/blob/f520dcd1d8b61cf35198ba7dc073b0fd633bbc4c/src/layout/group_invite.c#L27)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

LGTM but I'd also like to see `// TODO: OPTIMIZE` comments when we're doing something the easy way that is a good target for micro optimization.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/utox/utox/1074)*

**Comment by endoffile78** [comment]
:lgtm:

---

Reviewed 1 of 4 files at r9, 2 of 5 files at r10, 17 of 17 files at r11.
Review status: 17 files need 1 more reviewer, 3 open discussions remaining, 1 assignee LGTMs are missing, missing Admin LGTM at current revision. (waiting on @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, @undefined, and @undefined)

---



*Comments from [Reviewable](https://reviewable.io/reviews/utox/utox/1074#-:-LFG6Uo14a41LVT6I7vb:bnfp4nl)*

**Comment by redmanmale** [comment]
If no one mind it, I'd like to merge it. Are there any problems?

**Comment by avoidr** [comment]
I want to take a look, too. And push translations for Polish, too.

File: src/command_funcs.c [hunk 1]
```diff
@@ -52,7 +54,15 @@ bool slash_alias(void *object, char *arg, int arg_length) {
 bool slash_invite(void *object, char *arg, int UNUSED(arg_length)) {
     GROUPCHAT *g =  object;
     FRIEND *f = find_friend_by_name((uint8_t *)arg);
+
     if (f != NULL && f->online) {
+        size_t msg_length = UTOX_FRIEND_NAME_LENGTH(f) + SLEN(GROUP_MESSAGE_INVITE);
```
Comment (avoidr) [inline src/command_funcs.c]: Is the newline included in this calculation? It should be.
See here: https://github.com/uTox/uTox/pull/1273/commits

The variable should also be renamed, to `msg_size`, as in "buffer size" not "string length".

And don't use `snprintf`'s return value. Use `strnlen` instead.

File: src/flist.c [hunk 2]
```diff
@@ -221,8 +223,20 @@ static void drawitem(ITEM *i, int x, int y, int width) {
...
+            flist_draw_name(i, name_x, name_y, width, S(GROUP_CREATE), S(CURSOR_CLICK_RIGHT), SLEN(GROUP_CREATE),
+                            SLEN(CURSOR_CLICK_RIGHT), 1, (selected_item == i) ? COLOR_MAIN_TEXT : COLOR_LIST_TEXT);
+            break;
+        }
+
+        case ITEM_GROUP_INVITE: {
+            drawalpha(group_bitmap, avatar_x, y + ROSTER_AVATAR_TOP, default_w, default_w,
+                      selected_item == i ? COLOR_MAIN_TEXT : COLOR_LIST_TEXT);
+
+            size_t msg_length = SLEN(GROUP_INVITE) + 4;
```
Comment (avoidr) [inline src/flist.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/flist.c [hunk 3]
```diff
@@ -895,6 +941,14 @@ GROUPCHAT *flist_get_groupchat(void) {
     return NULL;
 }
 
+uint8_t flist_get_group_invite_id(void) {
+    if (flist_get_type() == ITEM_GROUP_INVITE) {
```
Comment (avoidr) [inline src/flist.c]: I would invert this and make it an early break implying error. As it is now, it implies that this condition being true is an exception which prevents this function from returning the correct value, which is `UINT8_MAX`.

File: src/flist.c [hunk 4]
```diff
@@ -1254,7 +1312,15 @@ bool flist_mup(void *UNUSED(n)) {
                 if (nitem->item == ITEM_GROUP) {
                     FRIEND *f = get_friend(selected_item->id_number);
                     GROUPCHAT *g = get_group(nitem->id_number);
+
                     if (f->online) {
+                        size_t msg_length = UTOX_FRIEND_NAME_LENGTH(f) + SLEN(GROUP_MESSAGE_INVITE);
```
Comment (avoidr) [inline src/flist.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/friend.c [hunk 5]
```diff
@@ -398,13 +398,12 @@ void friend_recvimage(FRIEND *f, NATIVE_IMAGE *native_image, uint16_t width, uin
 }
 
 void friend_notify_msg(FRIEND *f, const char *msg, size_t msg_length) {
-    char title[UTOX_FRIEND_NAME_LENGTH(f) + 25];
-
-    size_t title_length = snprintf((char *)title, UTOX_FRIEND_NAME_LENGTH(f) + 25, "uTox new message from %.*s",
-                                   (int)UTOX_FRIEND_NAME_LENGTH(f), UTOX_FRIEND_NAME(f));
+    size_t title_length = UTOX_FRIEND_NAME_LENGTH(f) + SLEN(FRIEND_MESSAGE_NEW);
```
Comment (avoidr) [inline src/friend.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/groups.c [hunk 6]
```diff
@@ -223,12 +223,9 @@ void group_peer_name_change(GROUPCHAT *g, uint32_t peer_id, const uint8_t *name,
     }
 
     if (peer->name_length) {
-        char old[TOX_MAX_NAME_LENGTH];
-        char msg[TOX_MAX_NAME_LENGTH];
-
-        memcpy(old, peer->name, peer->name_length);
-        size_t size = snprintf(msg, TOX_MAX_NAME_LENGTH, "<- has changed their name from %.*s",
-                               peer->name_length, old);
+        size_t size = peer->name_length + SLEN(GROUP_MESSAGE_CHANGE_NAME);
```
Comment (avoidr) [inline src/groups.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/groups.c [hunk 7]
```diff
@@ -344,11 +340,11 @@ void group_notify_msg(GROUPCHAT *g, const char *msg, size_t msg_length) {
         return;
     }
 
-    char title[g->name_length + 25];
-
-    size_t title_length = snprintf(title, g->name_length + 25, "uTox new message in %.*s", g->name_length, g->name);
+    size_t title_length = g->name_length + SLEN(GROUP_MESSAGE_NEW);
```
Comment (avoidr) [inline src/groups.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/layout/group.c [hunk 8]
```diff
@@ -56,8 +56,14 @@ static void draw_group(int x, int UNUSED(y), int UNUSED(w), int UNUSED(height))
         GROUP_PEER *peer = g->peer[i];
 
         if (peer && peer->name_length) {
-            char buf[TOX_MAX_NAME_LENGTH];
-            int  text_length = snprintf((char *)buf, TOX_MAX_NAME_LENGTH, "%.*s, ", (int)peer->name_length, peer->name);
+            uint8_t text_length = peer->name_length + 2;
```
Comment (avoidr) [inline src/layout/group.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/layout/group_invite.c [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+static void draw_group_invite(int x, int y, int UNUSED(w), int UNUSED(h)) {
+    const uint8_t invite_id = flist_get_group_invite_id();
+    const uint32_t friend_id = group_invite_get_friend_id(invite_id);
+    const FRIEND *f = get_friend(friend_id);
+
+    setcolor(COLOR_MAIN_TEXT);
+    setfont(FONT_SELF_NAME);
+
+    size_t label_length = SLEN(GROUP_INVITE) + 4;
```
Comment (avoidr) [inline src/layout/group_invite.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/utox.c [hunk 10]
```diff
@@ -632,7 +641,7 @@ void utox_message_dispatch(UTOX_MSG utox_msg_id, uint16_t param1, uint16_t param
                 g->source[param2] = g->source[g->peer_count];
             }
 
-            g->topic_length = snprintf((char *)g->topic, sizeof(g->topic), "%u users in chat", g->peer_count);
+            g->topic_length = snprintf((char *)g->topic, sizeof(g->topic), S(GROUP_STATUS), g->peer_count);
```
Comment (avoidr) [inline src/utox.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/utox.c [hunk 11]
```diff
@@ -653,7 +662,7 @@ void utox_message_dispatch(UTOX_MSG utox_msg_id, uint16_t param1, uint16_t param
                 return;
             }
 
-            g->topic_length = snprintf((char *)g->topic, sizeof(g->topic), "%u users in chat", g->peer_count);
+            g->topic_length = snprintf((char *)g->topic, sizeof(g->topic), S(GROUP_STATUS), g->peer_count);
```
Comment (avoidr) [inline src/utox.c]: Same as above; https://github.com/uTox/uTox/pull/1273/commits

File: src/group_invite.c [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+                             unsigned int samples, uint8_t channels, unsigned int sample_rate,
+                             void *userdata);
+
+bool group_invite_accept(Tox *tox, const uint8_t invite_id) {
+    if (invite_id >= MAX_GROUP_INVITES) {
+        LOG_ERR("Groupchat invite", "Invite number (%u) is out of bounds.", invite_id);
+        return false;
+    }
+
+    uint32_t group_id = UINT32_MAX;
```
Comment (avoidr) [inline src/group_invite.c]: Confusing initialisation.
Also consider deleting the two next empty lines, to show that it all belongs together.

**Comment by avoidr** [comment]
I don't like the current screen.

* When invited, the groupchat invitation appears on the friend list in a practically unnoticable fashion, even when you learned that it appears there, it is really hard to discover. It should be immediately obvious.

* The screen is inconsistent with other screens. It should probably look like the delete-friend screen; bold font and the buttons immediately below and next to each other. The buttons being at opposing ends of the window is not so cool. And the buttons being at the bottom of the window, with the query being at the top, is not that cool, either.

* The ignore button is shorter than its text.

**Comment by avoidr** [comment]
Also, there should be a setting for auto-accepting group invitations.

**Comment by avoidr** [comment]
Next I'm translating to German and Polish.

**Comment by avoidr** [comment]
@cebe, you can please look at my translation.

I am not sure about `LEAVE_GROUP`. I find "Gruppe verlassen" better, but then it's not consistent with `GROUP_CREATE`, if that is even so important. I find just "Gruppe" to be better.

**Pull request closed without merging**